python/lsst/daf/butler/_butler.py

```diff
@@ -1078,12 +1074,15 @@ def prune(self, refs: Iterable[DatasetRef], *,
                         # to ignore already-removed-from-datastore datasets
                         # anyway.
                         if self.datastore.exists(ref):
-                            self.datastore.remove(ref)
+                            self.datastore.trash(ref)
```

Comment:
I think we could now simplify a lot of this, removing the giant TODO comment block, and all the checks under the `if purge` blocks (which will be handled more rigorously by transaction rollback now).
 - Start a registry transaction context (I don't think the Datastore ones are useful here).
 - Expand datasets to components recursively.
 - Do the `if unstore` loop that calls `self.datastore.trash`.
 - Do the disassociate and and possibly the purge Registry operations.
 - End the transaction context, committing those changes (or raising if something went wrong).
 - Call `emptyTrash`.

I'd love to have someone else check my logic on that, of course, especially on whether it's safe if there are concurrent deletion operations and how it interacts with various kinds of composites in datastore (for virtual composites, are there ever records in `dataset_location`)?